components/scream/src/physics/p3/p3_functions_conservation_impl.hpp

```diff
@@ -0,0 +1,92 @@
+#ifndef P3_FUNCTIONS_CONSERVATION_IMPL_HPP
+#define P3_FUNCTIONS_CONSERVATION_IMPL_HPP
+
+#include "p3_functions.hpp"
+#include "p3_constants.hpp"
+
+namespace scream {
+namespace p3 {
+
+template<typename S, typename D>
+KOKKOS_FUNCTION
+void Functions<S,D>
+::cloud_water_conservation(const Spack& qc, const Spack& qcnuc,const Scalar dt,
+   Spack& qcaut, Spack& qcacc, Spack &qccol, Spack& qcheti, Spack& qcshd, Spack& qiberg, Spack& qisub, Spack& qidep)
+{
+  const auto sinks = (qcaut+qcacc+qccol+qcheti+qcshd+qiberg)*dt; // Sinks of cloud water
+  const auto sources = qc + (qcnuc)*dt; // Source of cloud water
+  Spack ratio;
+
+  constexpr Scalar qtendsmall = scream::p3::Constants<Scalar>::QTENDSMALL;
+  Smask enforce_conservation  = sinks > sources && sinks >= qtendsmall;  // determine if  conservation corrction is necessary
+  Smask nothing_todo = !enforce_conservation;
+
+  if (enforce_conservation.any()){
+    ratio.set(enforce_conservation, sources/sinks);
+    qcaut.set(enforce_conservation, qcaut*ratio);
+    qcacc.set(enforce_conservation, qcacc*ratio);
+    qccol.set(enforce_conservation, qccol*ratio);
+    qcheti.set(enforce_conservation, qcheti*ratio);
+    qcshd.set(enforce_conservation, qcshd*ratio);
+    qiberg.set(enforce_conservation, qiberg*ratio);
+  }
+
+  if(nothing_todo.any()){
+    ratio.set(nothing_todo, 1.0); // If not limiting sinks on qc then most likely did not run out of qc
+  }
+
+
+  //PMC: ratio is also frac of step w/ liq. thus we apply qiberg for
+  //"ratio" of timestep and vapor deposition and sublimation  for the
+  //remaining frac of the timestep.  Only limit if there will be cloud
+  // water to begin with.
+  enforce_conservation = sources > qtendsmall;
+  if (enforce_conservation.any()){
+    qidep.set(enforce_conservation, qidep*(1.0-ratio));
```

Comment:
Always use int literals where possible to avoid overpromotion which can lead to non-bfb issues.